// wfs-local/lib/policy.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const types_1 = require("./types");
/* tslint:disable:max-classes-per-file */
class CombinedPolicy {
    constructor(...all) {
        this._all = all;
    }
    comply(path, operation) {
        for (const one of this._all) {
            if (!one.comply(path, operation)) {
                return false;
            }
        }
        return true;
    }
}
exports.CombinedPolicy = CombinedPolicy;
class ReadOnlyPolicy {
    comply(path, operation) {
        if (operation === types_1.Operation.Read) {
            return true;
        }
        return false;
    }
}
exports.ReadOnlyPolicy = ReadOnlyPolicy;
class ForceRootPolicy {
    constructor(root) {
        this._root = root;
    }
    comply(path, operation) {
        if (path.indexOf(this._root) === 0) {
            return true;
        }
        return false;
    }
}
exports.ForceRootPolicy = ForceRootPolicy;
class AllowPolicy {
    comply(path, operation) { return true; }
}
exports.AllowPolicy = AllowPolicy;
class DenyPolicy {
    comply(path, operation) { return false; }
}
exports.DenyPolicy = DenyPolicy;
